### PR TITLE
Enforce 2fa with exclusion groups

### PR DESCRIPTION
**Pull request by jvillafanez**
## Description
Allow the admin to enforce 2fa usage. A 2fa app must be installed and enabled, otherwise this setting will be ignored and all the users will able to login normally.
The admin can also exclude certain groups from the 2fa enforcement. The users belonging to those groups won't have the 2fa enforced, but they can still use 2fa if they want.

## Related Issue
https://github.com/owncloud/enterprise/issues/5219

## How Has This Been Tested?
Manually tested along with the "twofactor_totp" app

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:

- [x] Code changes
- [x] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [ ] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

### CODE REVIEW

File: settings/Panels/Admin/Enforce2fa.php [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+class Enforce2fa implements ISettings {
+	/** @var IConfig */
+	protected $config;
+
+	public function __construct(IConfig $config) {
+		$this->config = $config;
+	}
+
+	public function getPriority() {
+		return 0;
```
Comment (mrow4a) [inline settings/Panels/Admin/Enforce2fa.php:35]: intentional?

File: lib/private/Authentication/TwoFactorAuth/Manager.php [hunk 2]
```diff
@@ -135,6 +160,10 @@ public function getProviders(IUser $user) {
 			}
 		}
 
+		if ($this->isTwoFactorEnforcedForUser($user)) {
```
Comment (mrow4a) [inline lib/private/Authentication/TwoFactorAuth/Manager.php:163]: add a comment why this is needed here. Note we check for this when calling `getProviders()`

**Review by mrow4a — COMMENTED** [review]
clarify

File: settings/templates/panels/admin/enforce2fa.php [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+?>
+<div class="section" id="2fa">
+	<h2 class="app-name"><?php p($l->t('Two-factor Authentication'));?></h2>
+	<p>
+		<em><?php p($l->t('This section requires a two-factor authentication app to be installed in ownCloud')); ?></em>
+	</p>
+	<h3><?php p($l->t('Enforce two-factor authentication')); ?></h3>
+	<p><?php p($l->t('Before enforcing the two-factor authentication, check the following requirements:')); ?></p>
+	<ul>
+		<li><?php p($l->t('At least a two-factor authentication app is installed and enabled in ownCloud.')); ?></li>
```
Comment (phil-davis) [inline settings/templates/panels/admin/enforce2fa.php:12]:
```suggestion
		<li><?php p($l->t('At least one two-factor authentication app is installed and enabled in ownCloud.')); ?></li>
```

File: settings/templates/panels/admin/enforce2fa.php [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+		<li><?php p($l->t('The users can setup at least a two-factor app from the challenge page. Some apps might not be prepared for this')); ?></li>
```
Comment (phil-davis) [inline settings/templates/panels/admin/enforce2fa.php:13]:
```suggestion
		<li><?php p($l->t('The users can setup at least one two-factor app from the challenge page. Some apps might not be prepared for this')); ?></li>
```

File: settings/templates/panels/admin/enforce2fa.php [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+	<br/>
+	<p>
+	<input type="checkbox" id="enforce_2fa" name="enforce_2fa" value="1" <?php if ($_['enforce2fa']) {
+		print_unescaped('checked="checked"');
+	}?> />
+	<label for="enforce_2fa"><?php p($l->t('Enforce two-factor authentication to all the users')); ?></label>
+	</p>
+	<br/>
+	<p>
+		<?php p($l->t('Exclude the following groups of enforcing the two-factor authentication')); ?>
```
Comment (phil-davis) [inline settings/templates/panels/admin/enforce2fa.php:25]:
```suggestion
		<?php p($l->t('Exclude the following groups from enforcing two-factor authentication')); ?>
```

File: settings/templates/panels/admin/enforce2fa.php [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+		<br />
+		<input name="enforce_2fa_excluded_groups" type="hidden" id="enforce_2fa_excluded_groups" value="<?php p($_['enforce2faExcludedGroups']) ?>" style="width: 400px">
+		<em>
+			<br />
+			<?php p($l->t('Users in these groups can use two-factor authentication on their own')); ?>
```
Comment (phil-davis) [inline settings/templates/panels/admin/enforce2fa.php:30]:
```suggestion
			<?php p($l->t('Note: Users in these groups can use two-factor authentication on their own')); ?>
```

should we put "Note" here to give a clue that 2FA will be optional for these users?

**Review by phil-davis — APPROVED** [review]

**Review by mrow4a — APPROVED** [review]

**Pull request merged**